### PR TITLE
[Util] Rework SimplifyGlobalAccesses to use DFX and support SCF

### PR DESCRIPTION
**Pull request by qedawkins**
This reimplements SimplifyGlobalAccesses using DFX to analyze the way different operations access globals, and then performs similar movement of global accessors (loads move up the CFG and stores move down) based on the analysis result. This adds proper tracking for nested regions and some support for invariant accessor hoisting from loop-like ops based on upstream interfaces.

NYI are
1. Motion through CF/branch based control flow. A slightly different addition to the analysis is probably needed to track when it is legal to hoist through a conditional branch, but this is currently assuming it will be easier/more common to handle earlier on at the SCF level.
2. Hoisting through function boundaries. Currently the propagation doesn't properly track accesses by functions after the initial analysis, but as noted in a comment this doesn't affect the outcome of any of the current patterns.
3. Hoisting out of scf.if/scf.while. Ifs are just NYI and require slightly different handling, and whiles are missing the interface function we use for scf.for to do the hoisting.

### CODE REVIEW

File: compiler/src/iree/compiler/Dialect/Util/Transforms/SimplifyGlobalAccesses.cpp [hunk 1]
```diff
@@ -78,240 +89,688 @@ static void hoistImmutableLoads(Region &region,
...
+        for (auto &[globalName, c] : getAssumedMultiSet()) {
+          llvm::dbgs() << "  " << globalName << ":  " << c << "\n";
+        }
+      } else {
+        llvm::dbgs() << ": ALL GLOBALS\n";
+      }
+    });
+  }
+
+  // The only operation kinds we need to update are
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/Util/Transforms/SimplifyGlobalAccesses.cpp]: (complete)

File: compiler/src/iree/compiler/Dialect/Util/Transforms/SimplifyGlobalAccesses.cpp [hunk 2]
```diff
@@ -78,240 +89,688 @@ static void hoistImmutableLoads(Region &region,
...
+};
+const char GlobalAccessor::ID = 0;
+
+class GlobalAccessorAnalysis {
+public:
+  explicit GlobalAccessorAnalysis(Operation *rootOp)
+      : explorer(rootOp, TraversalAction::RECURSE,
+                 OpPrintingFlags().elideLargeElementsAttrs().skipRegions()),
+        solver(explorer, allocator) {
+    // TODO: Use op traits or something to ignore the contents of executables
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/Util/Transforms/SimplifyGlobalAccesses.cpp]: I think IsolatedFromAbove should be enough to handle these and happen implicitly as the Explorer does its stuff - did you see it walking into these symbol table ops?

File: compiler/src/iree/compiler/Dialect/Util/Transforms/SimplifyGlobalAccesses.cpp [hunk 3]
```diff
@@ -78,240 +89,688 @@ static void hoistImmutableLoads(Region &region,
...
+    //   explorer.setOpAction<IREE::Stream::ExecutableOp>(TraversalAction::IGNORE);
+    //   explorer.setOpAction<IREE::HAL::ExecutableOp>(TraversalAction::IGNORE);
+    explorer.initialize();
+
+    assert(rootOp->getNumRegions() == 1 && "expected module-like root op");
+
+    // Populate the list of globals to optimize. This is primarily to force the
+    // order in which globals are processed to be deterministic, reducing minor
+    // ordering variations in the resulting IR.
+    rootOp->walk([&](IREE::Util::GlobalOpInterface global) {
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/Util/Transforms/SimplifyGlobalAccesses.cpp:285]: you can use explorer.forEachGlobal instead to avoid the full module walk

File: compiler/src/iree/compiler/Dialect/Util/Transforms/SimplifyGlobalAccesses.cpp [hunk 4]
```diff
@@ -78,240 +89,688 @@ static void hoistImmutableLoads(Region &region,
...
+  }
+
+  AsmState &getAsmState() { return solver.getAsmState(); }
+  Explorer &getExplorer() { return explorer; }
+
+  // Runs analysis and populates the state cache.
+  // May fail if analysis cannot be completed due to unsupported or unknown IR.
+  LogicalResult run() {
+    // Initialize all potential accessors.
+    explorer.walk([&](Operation *op) {
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/Util/Transforms/SimplifyGlobalAccesses.cpp:299]: use explorer.forEachGlobal to avoid the full module walk - it already should have all uses of global symbols in the `uses` list and you can use `isIndirect` to indicate a pessimistic state immediately

File: compiler/src/iree/compiler/Dialect/Util/Transforms/SimplifyGlobalAccesses.cpp [hunk 5]
```diff
@@ -78,240 +89,688 @@ static void hoistImmutableLoads(Region &region,
...
+    });
+
+    // Run solver to completion.
+    auto result = solver.run();
+    LLVM_DEBUG(solver.print(llvm::dbgs()));
+
+    // Populate the accessor map and other cached state for use when
+    // propagating direct accessors.
+    if (succeeded(result)) {
+      explorer.walk([&](Operation *op) {
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/Util/Transforms/SimplifyGlobalAccesses.cpp]: explorer.forEachGlobal here too to avoid the walk

File: compiler/src/iree/compiler/Dialect/Util/Transforms/SimplifyGlobalAccesses.cpp [hunk 6]
```diff
@@ -78,240 +89,688 @@ static void hoistImmutableLoads(Region &region,
...
+  // Move an operation right after the nearest blocking operation above |op|
+  // within its block. An operation is blocking if it *may* access the global
+  // referenced by |globalName|.
+  void moveOpUpInBlock(Operation *op, StringAttr globalName);
+  // Move an operation right before the nearest blocking operation below |op|
+  // within its block. An operation is blocking if it *may* access the global
+  // referenced by |globalName|.
+  void moveOpDownInBlock(Operation *op, StringAttr globalName);
+
+  StringAttr getGlobalName(Operation *op) {
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/Util/Transforms/SimplifyGlobalAccesses.cpp]: if you change to using GlobalInfo then you can just ask the globalInfo.op for its name

**Review by benvanik — CHANGES_REQUESTED** [review]
whoa, nice!

mostly just comments around preventing full module IR walks (especially now that this is not parallelized on function-like ops but running serialized over the entire module) - using GlobalInfo will ensure you use all the cached information without having to do that all yourself

File: compiler/src/iree/compiler/Dialect/Util/Transforms/SimplifyGlobalAccesses.cpp (hunk 4 above)
Comment (qedawkins) [reply compiler/src/iree/compiler/Dialect/Util/Transforms/SimplifyGlobalAccesses.cpp:299]: A comment and a couple questions: In addition to initializing the state for global.load/store, I'm also initializing it for things like external function calls, ops with Util_YieldPoint`, and indirect accessors. For external function calls I'm guessing I can just inspect the call graph, but I'm a bit unclear on how indirect accessors work nor how to handle yield points.

For indirect accessors, what I'm assuming now is that if I have an indirect accessor, I can't move any direct loads/stores past it, but I can still optimize around it, i.e.
```
// Some direct loads/stores I can optimize
util.global.load.indirect
// Some direct loads/stores I can optimize
```
I'm not checking whether the indirect load was loading/storing the same global as the nearby direct loads/stores (as I thought indirect more or less implied that we can't know which global it's accessing), but `GlobalInfo` seems to indicate that the explorer is able to do some reasoning about them? i.e. https://github.com/openxla/iree/blob/c27ed41cb17e8e8ce282cf9b838b0af1dd48c540/compiler/src/iree/compiler/Dialect/Util/Analysis/Explorer.h#L129

My expectation was that even if a global should be considered volatile due to indirect accesses, the indirect load and store ops are still the only way to interact with them, and thus we can still optimize other loads of that global. Is this understanding correct? Or do I have my understanding of indirect accessors mixed up (I kind of just pieced it together from code/comments).

If we say we'd rather not bother trying to ever optimize in the presence of indirect loads/stores that's fine and makes this easier, otherwise I'm not sure how to avoid the full walk because the explorer doesn't track these accessors.

My second question is w.r.t. yield points. I've been assuming we can't push loads/stores past, say, a `vm.break` because the user needs to be able to inspect the proper module state at that point in the code. As a result, I don't know how to avoid this walk for the same reason because I need to find all of the yield points throughout the module. Is there a reasonable way to populate a set of yield points on initialization of the explorer? Else I could let the initialization of yield points happen as an update step of the analysis, but that seems like it would introduce more IR walking than this.

Edit: If this wall of text is hard to parse I'm open to chatting to explain the problem I think I'm facing.

Comment (qedawkins) [reply compiler/src/iree/compiler/Dialect/Util/Transforms/SimplifyGlobalAccesses.cpp:299]: Should've read the code closer, I see that the explorer is just looking for globals where their address is taken, i.e. `GlobalAddressOpInterface`. So the explorer never sees the indirect loads/stores. I see that `initializeInverseCallGraph` still does a full module walk though, could we reuse that module walk to populate a list of yield points/indirect loads?

File: compiler/src/iree/compiler/Dialect/Util/Transforms/SimplifyGlobalAccesses.cpp (hunk 5 above)
Comment (qedawkins) [reply compiler/src/iree/compiler/Dialect/Util/Transforms/SimplifyGlobalAccesses.cpp]: I just iterated over the list of elements stored by the solver instead (as that's essentially what I wanted here).

File: compiler/src/iree/compiler/Dialect/Util/Transforms/SimplifyGlobalAccesses.cpp (hunk 6 above)
Comment (qedawkins) [reply compiler/src/iree/compiler/Dialect/Util/Transforms/SimplifyGlobalAccesses.cpp]: The propagation is rooted on `util.global.load`/`util.global.store`, and this is just supposed to be a helper for getting the name of the global from the attribute already on the op. I am already maintaining my own cache of Operation* -> Map<GlobalName, AccessCount> because I want to know, given an operation, does it access a particular global. This is what the analysis constructs and is easier to reuse directly. The `initializeOperation` that the solver internally calls back to when constructing a new element only gives me the operation and the solver itself, so it's much easier to just record the global name already tied to the `util.global.load`/`util.global.store` than to try to go and figure out which GlobalInfo it's tied to.

File: compiler/src/iree/compiler/Dialect/Util/Transforms/SimplifyGlobalAccesses.cpp (hunk 3 above)
Comment (qedawkins) [reply compiler/src/iree/compiler/Dialect/Util/Transforms/SimplifyGlobalAccesses.cpp:285]: `globalinfos` is a map so the order isn't deterministic. Is there another good way to get a deterministic order of the globals? I could go back to the symbol table the same way that `initializeGlobalInfos` does.

Comment (qedawkins) [reply compiler/src/iree/compiler/Dialect/Util/Transforms/SimplifyGlobalAccesses.cpp:285]: (the correctness of the pass doesn't depend on it, but I think it's good for the compiler's output to be deterministic).

File: compiler/src/iree/compiler/Dialect/Util/Analysis/Explorer.cpp [hunk 7]
```diff
@@ -112,6 +118,7 @@ void Explorer::initializeGlobalInfos() {
     } else {
       globalInfo.uses.push_back(use.getUser());
     }
+    orderedGlobalInfos.push_back(&globalInfo);
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Dialect/Util/Analysis/Explorer.cpp]: ~~W.R.T the deterministic processing of globals I did this, but I don't think it's good that it's tracking every global info twice for all other users of Explorer that don't care about this.~~ reverted

**Comment by benvanik** [comment]
probably worth a quick sync - RE indirect: we don't really support them today anywhere else and disable optimizations in their presence, so that's fine to do here

**Review by benvanik — APPROVED** [review]

File: compiler/src/iree/compiler/Dialect/Util/Transforms/SimplifyGlobalAccesses.cpp [hunk 8]
```diff
@@ -78,240 +103,737 @@ static void hoistImmutableLoads(Region &region,
...
+    // has no effect on the current reached fixedpoint given that accessors
+    // are only ever removed one at a time from a pair within the same region.
+    while (parent) {
+      decrementAccessorCount(globalName, parent);
+      parent = parent->getParentOp();
+    }
+    op->erase();
+  }
+
+  bool operationDoesAccessGlobal(Operation *op, StringAttr globalName) const {
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Dialect/Util/Transforms/SimplifyGlobalAccesses.cpp:426]: Almost all of the non-analysis time is spent here, just querying whether or not an operation accesses a particular global. This is all cached analysis state so I'm not entirely sure how to make this much faster; maybe there are some better data structures than DenseMap. Otherwise, there might be a way to reorganize the processing of ops that saves some traversal of ops surrounding global loads/stores, but my attempts to improve this haven't worked at least.

**Review by qedawkins — COMMENTED** [review]
I tried to do some local benchmarking of the changes here, and found that this version of the pass has a noticeable impact on performance (the full compile stats also seem to suggest this, but it wasn't enough to flag a regression I guess).

I attempted to rework this pass a bit to make it so that it could nest on functions as well (at the cost of some optimizations) but it wasn't faster and it felt like I was bending the Explorer to do something it wasn't supposed to so I undid it.

File: compiler/src/iree/compiler/Dialect/Util/Transforms/SimplifyGlobalAccesses.cpp (hunk 8 above)
Comment (benvanik) [reply compiler/src/iree/compiler/Dialect/Util/Transforms/SimplifyGlobalAccesses.cpp:426]: This is inefficient as it's doing doing 4 lookups for each op and then two lookups for each global within that op.
You _must_ think about access patterns for things like maps/sets/etc (and even vectors, as there's debug overhead in doing vector access). Instead of having a globalAccessors and accessorCounts and then a map within the accessor counts you could just have accessorCounts keyed on `std::tuple<Operation *, StringAttr>` - then you have a single lookup to perform:
```c++
auto it = accessorCounts.find(std::make_tuple(op, globalName));
if (it == accessorCounts.end()) return false;  // does not access
return it->second != 0;
```

This kind of thing applies in several places here (like the following function) - the rule to always stick to is that you get one single access of a map and you need to make it count - so don't use contains and at, as find lets you do it in one access, and don't use multiple ats as you could just do it once and reuse that result.

Algorithmically there's probably something different that could be done. For example, if this is n^2 (each op checking every other op) and the reason the perf is bad is because then you have these multiple map lookups in your inner loop you could instead invert the flow to run through the function and calculate per global all ops that access it then run down that list per global - in the common case there will be one or two ops there and average complexity goes down to n (the single walk over all ops to compute the sets). Then when solving the worst case (a billion accesses to the same global) you can have better data structures as you know that all ops in that per global set both access a global and access the one under inspection so you don't need to perform any map lookups and can instead do something linear and end up still at n.

**Review by benvanik — CHANGES_REQUESTED** [review]

**Comment by qedawkins** [comment]
Closing as stale, a lot has changed since this was sent.
